### PR TITLE
Use doccoms from signature and implementation files for quickinfo tooltips

### PR DESCRIPTION
**Pull request by cloudRoutine**
The current way that documentation comments behave with quickinfo tooltips is rather odd and has been a pet peeve of mine for a long time.

When you add doccoms in the signature they don't show up in the corresponding implementation  

If you open a different implementation file it uses the doccoms from the sig and overwrites any doccoms you may have added in the implementation file

Now it shows doccoms from the sig in its corresponding implementation file, and if you add additional doccoms in the implementation the tooltip will show both of them.

The doccoms from the implementation files are on the bottom because these are merely supplementary and will not be included in the xml documentation created for the project.

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
Awesome. What about showing a warning of some sort if documentation in sig and implementation files differ (and a quick fix to synchronize them in either way ;) ?

**Comment by cloudRoutine** [comment]
The doccoms in the implementation files are more like "design time" notes, things you want to show up in other implementer's tooltips but prefer not to expose publicly like the doccoms in the signature file.

I don't think synchronization adds anything useful as the doccoms from the signature will always show in the tooltips. If anything you'd want to delete the doccoms in the implementation file. Making them the same would just duplicate the content.

File: vsintegration/src/FSharp.Editor/Common/Pervasive.fs [hunk 1]
```diff
@@ -199,6 +223,41 @@ module Async =
             }
         async { return! agent.PostAndAsyncReply id }
 
+
+type Async with
+    /// Better implementation of Async.AwaitTask that correctly passes the exception of a failed task to the async mechanism
+    static member AwaitTaskCorrect (task:Task) : unit Async =
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/Pervasive.fs]: ``Async<unit>`` please.

File: vsintegration/src/FSharp.Editor/Common/Pervasive.fs [hunk 2]
```diff
@@ -199,6 +223,41 @@ module Async =
...
+                    if e.InnerExceptions.Count = 1 then 
+                        exceptionCont e.InnerExceptions.[0]
+                    else exceptionCont e
+                elif task.IsCanceled then
+                    exceptionCont(TaskCanceledException ())
+                else successCont ())
+            |> ignore)
+
+    /// Better implementation of Async.AwaitTask that correctly passes the exception of a failed task to the async mechanism
+    static member AwaitTaskCorrect (task:'T Task) : 'T Async =
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/Pervasive.fs]: ``Task<'T>`` and ``Async<'T>`` please

File: vsintegration/src/FSharp.Editor/Common/Pervasive.fs [hunk 3]
```diff
@@ -199,6 +223,41 @@ module Async =
             }
         async { return! agent.PostAndAsyncReply id }
 
+
+type Async with
+    /// Better implementation of Async.AwaitTask that correctly passes the exception of a failed task to the async mechanism
+    static member AwaitTaskCorrect (task:Task) : unit Async =
+        Async.FromContinuations (fun (successCont,exceptionCont,_cancelCont) ->
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/Pervasive.fs]: If this ``AwaitTask`` is better, then we should start an RFC to provide a version of it in FSharp.Core please?

File: vsintegration/src/FSharp.Editor/QuickInfo/QuickInfoProvider.fs [hunk 4]
```diff
@@ -11,76 +11,243 @@ open Microsoft.CodeAnalysis
...
+            
+        // it is necessary to retrieve the backup tooltip info because this acquires
+        // the textSpan designating where we want the tooltip to appear.
+        let! backupTooltipInfo 
+            & (_, targetTextSpan, _, _) = getTargetSymbolTooltip()
+   
+        let tooltipInfoPair = asyncMaybe {
+            match findSigDeclarationResult with 
+            | FSharpFindDeclResult.DeclNotFound _failReason -> 
+                return (None , Some backupTooltipInfo)
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfo/QuickInfoProvider.fs]: No space before comma please

File: vsintegration/src/FSharp.Editor/QuickInfo/QuickInfoProvider.fs [hunk 5]
```diff
@@ -11,76 +11,243 @@ open Microsoft.CodeAnalysis
...
+                    let! sigTooltipInfo = getTooltipFromRange declRange
+                    // if the target was declared in a signature file, and the current file
+                    // is not the corresponding module implementation file for that signature,
+                    // the doccoms from the signature will overwrite any doccoms that might be 
+                    // present on the definition/implementation
+
+                    let! findImplDefinitionResult = 
+                        checkFileResults.GetDeclarationLocationAlternate
+                            ( idRange.StartLine, idRange.EndColumn, lineText
+                            , lexerSymbol.FullIsland, preferFlag=false) |> liftAsync
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/QuickInfo/QuickInfoProvider.fs]: Put comma at end-of-line please, not start of line

File: vsintegration/src/FSharp.Editor/Common/Pervasive.fs [hunk 6]
```diff
@@ -20,6 +23,27 @@ module String =
...
+        with _ -> path
+
+    static member GetFileNameSafe path =
+        try Path.GetFileName path
+        with _ -> path
+
+
+/// Load times used to reset type checking properly on script/project load/unload. It just has to be unique for each project load/reload.
+/// Not yet sure if this works for scripts.
+let fakeDateTimeRepresentingTimeLoaded x = DateTime(abs (int64 (match x with null -> 0 | _ -> x.GetHashCode())) % 103231L)
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/Common/Pervasive.fs]: Why is this added in this PR?  It doesn't seem to be mentioned elsewhere in the PR?

**Comment by dsyme** [comment]
@cloudRoutine An alternative is to fix this within the compiler itself, so that the specification of documentation generation changes to combine documentation from signature and implementation. If implementation documentation is empty then it is uncontroversial to use signature documentation. However frequently documentation is duplicated in signature and implementation, which raises questions about what to do in that case.

**Comment by KevinRansom** [comment]
build error:
````
"D:\j\workspace\release_ci_pa---3f142ccc\build-everything.proj" (default target) (1) ->
"D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\fsharp-vsintegration-unittests-build.proj" (Build target) (38) ->
"D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\VisualFSharp.Unittests.fsproj" (Build target) (40) ->
(CoreCompile target) -> 
  D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\QuickInfoProviderTests.fs(102,37): error FS0039: The field, constructor or member 'ProvideQuickInfo' is not defined. [D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\VisualFSharp.Unittests.fsproj]

    13 Warning(s)
    1 Error(s)
````

File: vsintegration/src/FSharp.Editor/Common/Pervasive.fs (hunk 3 above)
Comment (cloudRoutine) [inline vsintegration/src/FSharp.Editor/Common/Pervasive.fs]: @eiriktsarpalis wrote it

**Comment by cloudRoutine** [comment]
@KevinRansom an issue I've run into with testing new features built around the Roslyn functionality is that there's no easy way to create a workspace to test it out with. Since the default workspace implementations reject all code that isn't C# or vb funtions/methods that rely on `Document` or `Project` connections inside of a solution are really hard to test.

**Comment by cloudRoutine** [comment]
@dsyme maybe that can come later, but I'm trying to address a more immediate concern I've had while working on this repo in particular due to the extensive use of `.fsi` files.

> However frequently documentation is duplicated in signature and implementation, which raises questions about what to do in that case.

this is a crude workaround due to inadequate tooling support and I think trying to account for it is going to end up in a lot of generated documentation suddenly becoming very messy. Trying to establish a good heuristic for when doccoms are similar enough to be duplicates seems like an invitation for kludges.

**Comment by cloudRoutine** [comment]
Is there anything else I need to do for this?

**Comment by cloudRoutine** [comment]
this might even encourage people to spend more time writing documentation in signature files since they'd actually get the benefit of it in their implementation files. the giant gross xml doccoms won't have to pollute their code.

**Comment by cartermp** [comment]
@cloudRoutine agreed

**Comment by cloudRoutine** [comment]
@KevinRansom this'll be pretty useful when working on the parts of the compiler where there's extensive documentation in the signatures but barely any in the implementation files. So hopefully it can get merged soon.

**Comment by cloudRoutine** [comment]
@KevinRansom all green and resolved here too

**Comment by cloudRoutine** [comment]
integrated into https://github.com/Microsoft/visualfsharp/pull/2683
